js/liqui.js

```diff
@@ -617,21 +626,39 @@ module.exports = class liqui extends Exchange {
         return { 'url': url, 'method': method, 'body': body, 'headers': headers };
     }
 
-    async request (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
-        let response = await this.fetch2 (path, api, method, params, headers, body);
+    handleErrors (httpCode, reason, url, method, headers, body) {
+        if (httpCode !== 200)
+            return; // resort to default error handler
+        let response = undefined;
+        if ((body[0] === '{') || (body[0] === '[')) {
+            response = JSON.parse (body);
+        } else {
+            // if not a JSON response
+            throw new ExchangeError (this.id + ' returned a non-JSON reply: ' + body);
+        }
         if ('success' in response) {
             if (!response['success']) {
```

Comment:
@mkutny , this line is a problem that @Deathamns was talking about... 
- there are 5 exchanges derived from Liqui, Liqui returns boolean or integer codes, others return string-booleans "true/false", which cannot be ported blindly, because in Python a stringified boolean will be in uppercase ("True/False")... In Liqui+JS it will work anyway, but it will most likely fail to work in derived+Python.

Reply:
It this the right line? It's a JSON check...

Reply:
Yes, the `!response['success']` check fails if (and only if, don't miss this important note) we are talking of an exchange derived from Liqui in Python that would return boolean strings instead of booleans for the `'success'` key ("true/false") like so:

```
>>> not ""
True
>>> not "true"
False
>>> not "false"
False
>>> 
```

Reply:
1) Ok, I see. I know already about `safe*` methods, but I never thought python's json is incompatible with js. Actually, json is a standard in the first place.

2) At the same time, you need to admit that this chunk was always there, I didn't touch it. It's not to blame someone, just to stress how easily these things could be overlooked.

3) `safe*` methods won't help anyway... I see something like that could be required:
```
if ('success' in response)
  let success = response['success']
  let unbielievebleSuccess = success === true || success === 'true' || success === 'True' || success > 0
```
4) What if we write something like `json.safeParse()` for python that's inspects every property and changes `true/false` for `True/False`? How do python people usually handle json?

Reply:
> Ok, I see. I know already about safe* methods, but I never thought python's json is incompatible with js. Actually, json is a standard in the first place.

> What if we write something like json.safeParse() for python that's inspects every property and changes true/false for True/False? How do python people usually handle json?

Nope, you didn't get it entirely. JSON is fine. It is standard in all languages. I am talking of a derived exchange (from Liqui), that is sending JSON-strings ("true/false") instead of JSON-bools (true/false), whereas Liqui sends them as booleans. So, it's not a Python vs non-Python difference, it's a Liqui vs liqui-child difference. And that difference breaks Python in its turn. So, Python is a consequence here, and the reason is the changing type of the value. This is what I was trying to tell in my previous message. I think we will be ok with `safeCoerceToBoolean()` or something like that (a variation of point 3).

Reply:
Ok, let me iterate one more time, maybe I don't understand something.

1. An exchange sends "true/false" value, instead of JSON's `true/false`.
2. Standard Pythons `json` (I just checked) doesn't not convert native JSON bolean types to native Python boolean types.
3. If I were a Python guy I would invent some "almost official" `json` importer alternative which would make such a trivial conversion, so I believe there is such converter available. Or maybe they have something like `JSON.parse(revival)`?
4. If we use this alternative for `Exchange.py` it would ease our lives, right?

Reply:
No, no, no ) 

All wrong starting from 1 )

Here's what I'm trying to tell you, same in all languages JS and Python (and PHP, not shown on screenshot below):

<img width="602" alt="screen shot 2018-01-15 at 19 47 01" src="https://user-images.githubusercontent.com/1294454/34953145-07a11cfc-fa2d-11e7-9241-34a2469954a3.png">

If an exchange sends true/false – we're fine. But if the exchange sends "true/false" and we are doing if (!"true/false") – we're doomed in all languages. Python is just more sensitive and crashes where JS and PHP survives.

Reply:
Anyway, this should be solvable with something similar to what you proposed in point 3.

Reply:
Ok, I love code snippets too:

```
> function safeBoolean(key, value) {
... const map = {'true': true, 'false': false, 'True': true, 'False': false };
... return (value in map) ? map[value] : value;
... }
undefined
> JSON.parse('{"foo":"true", "bar":"false", "baz":"True", "boo": "False"}', safeBoolean)
{ foo: true, bar: false, baz: true, boo: false }
```

It's not that I'm lazy to code `unbreakableSuccess` approach but my point is that such unsafe places are quite easy to miss. This one has been here for ages and I believe there some more unsafe places in other implementations as well.

But if we code this `JSON.parse(data, safeBoolean)` at the base class level it can save us from problems to some extent.

Reply:
Just to clarify, I didn't mean to blame you in any way, it's mostly me commenting my own older code, so, yes, entirely my fault ) But we will fix it anyway )

Reply:
I'm not about blaming either! We'll surely fix it, I just want to consider two alternative approaches:

1) `unbreakableSucces = success === true || success === 'true' || success === 'True' || ...`
2) `JSON.parse (data, safeBoolean)`

I'm really missing your opinion on the `safeBoolean` code snippet above.

Reply:
> opinion on the safeBoolean code snippet above

Oh, ok ) Well, from my crippled explanation, I tried to show, that the problem isn't related to the JSON-standard in itself at all. The problem appears on a later stage, when we check the deserialized value. It can be either:
1. `response['success'] === true` (a boolen, as with Liqui  ← Exchange)
2. `response['success'] === 'true'` (a string, as with Yobit ← Liqui ← Exchange)

Now, on point 1: we can only write these types of conditionals like so:
```JavaScript
if (response['success'])
// or
if (!response['success'])
```
We can't write `if (response['success'] === true)`, without altering the transpiler further, because Python PEP8 forbids equality-comparisons to booleans, it only permits 'is' or 'not' keywords (boolean syntax/semantics is slightly different there). So we can either fix the transpiler for it (make it reduce the `=== true` and `!== true` clauses to a basic `is/not`), or we can do it on a per-exchange base.

On point 2: any language will break on `if (!stringValue)` vs `if (stringValue)`, so when the exchange returns a string, this is not about Python, JSON or whatever, this is about negating a string, which is wrong, and we have to do a strict-comparison.

So, the solution is to combine 1+2 in Liqui. An alternative solution would be to redefine the error handler in all exchanges derived from Liqui. Did I just make any sense?

Reply:
Actually, what you say makes a lot of sense.

I submitted a patch, you may check if you like it.